crypto/crypto.go

```diff
@@ -0,0 +1,153 @@
+		mod.InitConsensusModule(mods, cfg)
+	}
+}
+
+// CreatePartialCert signs a single block and returns the partial certificate.
+func (c crypto) CreatePartialCert(block *consensus.Block) (cert consensus.PartialCert, err error) {
+	sig, err := c.Sign(block.ToBytes())
+	if err != nil {
+		return consensus.PartialCert{}, err
+	}
+	return consensus.NewPartialCert(sig, block.Hash()), nil
+}
+
+// CreateQuorumCert creates a quorum certificate from a list of partial certificates.
+func (c crypto) CreateQuorumCert(block *consensus.Block, signatures []consensus.PartialCert) (cert consensus.QuorumCert, err error) {
+	// genesis QC is always valid.
+	if block.Hash() == consensus.GetGenesis().Hash() {
+		return consensus.NewQuorumCert(nil, 0, consensus.GetGenesis().Hash()), nil
+	}
+	sigs := make([]consensus.QuorumSignature, 0, len(signatures))
+	for _, sig := range signatures {
+		sigs = append(sigs, sig.Signature())
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.QuorumCert{}, err
+	}
+	return consensus.NewQuorumCert(sig, block.View(), block.Hash()), nil
+}
+
+// CreateTimeoutCert creates a timeout certificate from a list of timeout messages.
+func (c crypto) CreateTimeoutCert(view consensus.View, timeouts []consensus.TimeoutMsg) (cert consensus.TimeoutCert, err error) {
+	// view 0 is always valid.
+	if view == 0 {
+		return consensus.NewTimeoutCert(nil, 0), nil
+	}
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		sigs = append(sigs, timeout.ViewSignature)
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.TimeoutCert{}, err
+	}
+	return consensus.NewTimeoutCert(sig, view), nil
+}
+
+// CreateAggregateQC creates an AggregateQC from the given timeout messages.
+func (c crypto) CreateAggregateQC(view consensus.View, timeouts []consensus.TimeoutMsg) (aggQC consensus.AggregateQC, err error) {
+	qcs := make(map[hotstuff.ID]consensus.QuorumCert)
+	sigs := make([]consensus.QuorumSignature, 0, len(timeouts))
+	for _, timeout := range timeouts {
+		if qc, ok := timeout.SyncInfo.QC(); ok {
+			qcs[timeout.ID] = qc
+		}
+		if timeout.MsgSignature != nil {
+			sigs = append(sigs, timeout.MsgSignature)
+		}
+	}
+	sig, err := c.Combine(sigs...)
+	if err != nil {
+		return consensus.AggregateQC{}, err
+	}
+	return consensus.NewAggregateQC(qcs, sig, view), nil
+}
+
+// VerifyPartialCert verifies a single partial certificate.
+func (c crypto) VerifyPartialCert(cert consensus.PartialCert) bool {
+	block, ok := c.mods.BlockChain().Get(cert.BlockHash())
+	if !ok {
+		return false
+	}
+	return c.Verify(cert.Signature(), block.ToBytes())
+}
+
+// VerifyQuorumCert verifies a quorum certificate.
+func (c crypto) VerifyQuorumCert(qc consensus.QuorumCert) bool {
+	// genesis QC is always valid.
+	if qc.BlockHash() == consensus.GetGenesis().Hash() {
+		return true
+	}
+	if qc.Signature().Participants().Len() < c.mods.Configuration().QuorumSize() {
```

Comment:
Similar to my [comment above](https://github.com/relab/hotstuff/pull/36/files#r888374558), the quorum size or Configuration could be passed in, which would decouple the crypto package from the module system.

At least my initial reaction was that this may cause problems due to tight coupling.

Reply:
I think that passing in the quorum size or configuration complicates the interface. I don't think the Crypto module's dependency on the BlockChain module is a problem.